util/git/git.go

```diff
@@ -43,20 +43,24 @@ func IsTruncatedCommitSHA(sha string) bool {
 
 // NormalizeGitURL normalizes a git URL for lookup and storage
 func NormalizeGitURL(repo string) string {
-	// preprocess
 	repo = strings.TrimSpace(repo)
-	repo = ensureSuffix(repo, ".git")
 	if IsSSHURL(repo) {
 		repo = ensurePrefix(repo, "ssh://")
 	}
-
-	// process
 	repoURL, err := url.Parse(repo)
 	if err != nil {
 		return ""
 	}
-
-	// postprocess
+	hostname := repoURL.Hostname()
+	if !strings.HasSuffix(hostname, "dev.azure.com") && !strings.HasSuffix(hostname, "visualstudio.com") {
+		// NOTE: not all git services support `.git` appended to their URLs (e.g. azure) so this
+		// normalization is not quite safe.
+		repo = ensureSuffix(repo, ".git")
+		repoURL, err = url.Parse(repo)
```

Comment:
Why are we re-parsing after adding the `.git` suffix? I'm having some difficulty coming up with a URL path that becomes invalid as a URL once a suffix is added.

Reply:
It's not about URL parsing/validation. We need to append the git suffix conditionally, only after we determine the hostname/domain is not azure. So the first parse was used to allow us to extract the domain. The second is done after we determine it's safe to append the suffix.